File: src/haruhiloop_cli/engine.py
from __future__ import annotations

from dataclasses import dataclass

from haruhiloop_cli.models import Action, Ending, GameState, StepRecord, TIMESLOTS, clamp
from haruhiloop_cli import action_flavor_zh, i18n, rules
from haruhiloop_cli.mutator import build_mutator, validate_profile
from haruhiloop_cli.systems.closed_space import evaluate_closed_space_stage
from haruhiloop_cli.systems.crew import apply_crew_sync
from haruhiloop_cli.systems.homework import apply_homework_progress, maybe_trigger_homework_pressure
from haruhiloop_cli.systems.memory import apply_memory_residue, apply_residue_decay, inject_residue_bonus


@dataclass(slots=True)
class StepResult:
    state: GameState
    record: StepRecord
    action: Action
    events: list[str]
    ending: Ending | None


class GameEngine:
    def create_new_state(
        self,
        run_id: str,
        *,
        mutator_mode: str = "ai",
        random_seed: int | None = None,
        ai_temperature: float = 0.7,
    ) -> GameState:
        state = GameState(
            run_id=run_id,
            mutator_mode=mutator_mode,
            random_seed=random_seed,
            ai_temperature=ai_temperature,
        )
        state.worldline_mutation_profile = self._compute_profile(state)
        return state

    def available_actions(self, state: GameState) -> list[Action]:
        return rules.get_available_actions(state)

    def step(self, state: GameState, action_id: str, step_number: int) -> StepResult:
        if state.is_finished:
            label = i18n.format_ending_summary(state.ending_id)
            raise ValueError(f"本局已结束，结局：{label}")

        action = rules.ACTIONS.get(action_id)
        if action is None:
            raise ValueError(f"未知动作：{action_id}")

        before = state.snapshot()
        profile = state.worldline_mutation_profile or self._compute_profile(state)
        sat_factor = profile.get("satisfaction_factor", 1.0)
        stab_factor = profile.get("stability_factor", 1.0)
        clue_factor = profile.get("clue_factor", 1.0)

        state.satisfaction = clamp(state.satisfaction + round(action.delta_satisfaction * sat_factor))
        state.stability = clamp(state.stability + round(action.delta_stability * stab_factor))
        state.clue_points = clamp(state.clue_points + round(action.delta_clue_points * clue_factor))
        state.nagato_fatigue = clamp(state.nagato_fatigue + action.delta_nagato_fatigue)
        state.flags.update(action.add_flags)
        self._update_streak(state, action_id)
        state.action_counts[action_id] = state.action_counts.get(action_id, 0) + 1
        for category_id in rules.action_categories_for(action_id):
            state.category_counts[category_id] = state.category_counts.get(category_id, 0) + 1
        action_flavor = action_flavor_zh.pick_action_flavor(state, action_id, step_number)
        state.recent_actions.append(action_id)
        state.recent_actions = state.recent_actions[-8:]
        state.worldline_shift += abs(action.delta_satisfaction) + abs(action.delta_clue_points)

        triggered = []
        triggered.extend(apply_homework_progress(state, action_id))
        triggered.extend(apply_crew_sync(state, action_id))
        triggered.extend(rules.evaluate_events(state, action))
        triggered.extend(evaluate_closed_space_stage(state, action_id))
        triggered.extend(maybe_trigger_homework_pressure(state))

        inject_residue_bonus(state)
        event_labels: list[str] = []
        for event in triggered:
            state.satisfaction = clamp(state.satisfaction + event.delta_satisfaction)
            state.stability = clamp(state.stability + event.delta_stability)
            state.clue_points = clamp(state.clue_points + event.delta_clue_points)
            state.flags.update(event.add_flags)
            event_labels.append(i18n.format_event_line(event))

        ending = rules.evaluate_ending(state)
        if ending is not None:
            state.ending_id = ending.ending_id
            state.ending_title = ending.title
            state.ending_epilogue = ending.description

        self._advance_time(state)

        record = StepRecord(
            step_number=step_number,
            day=before["day"],
            timeslot=before["timeslot"],
            action_id=action.action_id,
            action_label=action.label,
            before=before,
            after=state.snapshot(),
            events=event_labels,
            mutation_profile=dict(profile),
            ending_id=state.ending_id,
            action_flavor=action_flavor,
        )
        return StepResult(state=state, record=record, action=action, events=event_labels, ending=ending)

    def _compute_profile(self, state: GameState) -> dict[str, float]:
        seed = state.random_seed
        if state.mutator_mode == "ai" and seed is not None:
            seed = seed + state.day * 97 + state.loop_count * 193
        mutator = build_mutator(
            state.mutator_mode,
            seed=seed,
            temperature=state.ai_temperature,
        )
        try:
            raw_profile = mutator.mutate(state)
            return validate_profile(raw_profile)
        except Exception:
            fallback = build_mutator("deterministic")
            return validate_profile(fallback.mutate(state))

    @staticmethod
    def _update_streak(state: GameState, action_id: str) -> None:
        if state.previous_action == action_id:
            state.current_action_streak += 1
        else:
            state.current_action_streak = 0
        state.previous_action = action_id

    def _advance_time(self, state: GameState) -> None:
        if state.timeslot_index == len(TIMESLOTS) - 1:
            apply_memory_residue(state)
            state.timeslot_index = 0
            state.day += 1
            state.loop_count += 1
            # Slow drift emphasizes infinite-loop pressure.
            state.satisfaction = clamp(state.satisfaction - 1)
            state.stability = clamp(state.stability - 1)
            apply_residue_decay(state)
            state.worldline_mutation_profile = self._compute_profile(state)
        else:
            state.timeslot_index += 1

File: web/src/domain/engine.ts
// 与 src/haruhiloop_cli/engine.py 1:1 对齐：步骤推进、时间推进、扰动 profile 计算。

import {
  Ending,
  GameState,
  MutationProfile,
  Scene,
  SceneChoice,
  StepCommand,
  StepRecord,
  TIMESLOTS,
  createGameState,
  isFinished,
  snapshot,
  timeslotOf,
} from "./models";
import {
  findChoice,
  getAvailableChoices,
  getAvailableScenes,
  getScene,
} from "./rules/scenes";
import { evaluateEvents } from "./rules/events";
import { evaluateEnding } from "./rules/endings";
import { applyHomeworkProgress, maybeTriggerHomeworkPressure } from "./systems/homework";
import { applyCrewSync } from "./systems/crew";
import { evaluateClosedSpaceStage } from "./systems/closed_space";
import {
  applyMemoryResidue,
  applyResidueDecay,
  injectResidueBonus,
} from "./systems/memory";
import { buildMutator, validateProfile } from "./mutator";
import { pickChoiceFlavor } from "./action_flavor";
import { formatEventLine, formatEndingSummary } from "../narrative/i18n";
import { clamp } from "../infrastructure/clamp";

export interface StepResult {
  state: GameState;
  record: StepRecord;
  scene: Scene;
  choice: SceneChoice;
  events: string[];
  ending: Ending | null;
}

export class GameEngine {
  createNewState(
    runId: string,
    options: {
      mutatorMode?: "ai" | "deterministic";
      randomSeed?: number | null;
      aiTemperature?: number;
    } = {},
  ): GameState {
    const state = createGameState(runId, {
      mutator_mode: options.mutatorMode ?? "ai",
      random_seed: options.randomSeed ?? null,
      ai_temperature: options.aiTemperature ?? 0.7,
    });
    state.worldline_mutation_profile = this._computeProfile(state);
    return state;
  }

  availableScenes(state: GameState): Scene[] {
    return getAvailableScenes(state);
  }

  availableChoices(state: GameState, sceneId: string): SceneChoice[] {
    return getAvailableChoices(state, sceneId);
  }

  step(state: GameState, command: StepCommand, stepNumber: number): StepResult {
    if (isFinished(state)) {
      const label = formatEndingSummary(state.ending_id);
      throw new Error(`本局已结束，结局：${label}`);
    }

    const scene = getScene(command.scene_id);
    if (!scene) throw new Error(`未知场景：${command.scene_id}`);
    if (!this.availableScenes(state).some((s) => s.scene_id === scene.scene_id)) {
      throw new Error(`当前时段不可进入场景：${command.scene_id}`);
    }
    const choice = findChoice(command.scene_id, command.choice_id);
    if (!choice) {
      throw new Error(`场景 ${command.scene_id} 下无选项：${command.choice_id}`);
    }

    const before = snapshot(state);
    const profile: MutationProfile =
      state.worldline_mutation_profile ?? this._computeProfile(state);
    const satFactor = profile.satisfaction_factor ?? 1.0;
    const stabFactor = profile.stability_factor ?? 1.0;
    const clueFactor = profile.clue_factor ?? 1.0;

    state.satisfaction = clamp(
      state.satisfaction + Math.round(choice.delta_satisfaction * satFactor),
    );
    state.stability = clamp(
      state.stability + Math.round(choice.delta_stability * stabFactor),
    );
    state.clue_points = clamp(
      state.clue_points + Math.round(choice.delta_clue_points * clueFactor),
    );
    state.nagato_fatigue = clamp(state.nagato_fatigue + choice.delta_nagato_fatigue);
    for (const f of choice.add_flags) state.flags.add(f);
    this._updateStreak(state, choice.choice_id);
    state.scene_choice_counts[choice.choice_id] =
      (state.scene_choice_counts[choice.choice_id] ?? 0) + 1;
    this._applyRouteUpdates(state, choice);

    const actionFlavor = pickChoiceFlavor(state, scene.scene_id, choice.choice_id, stepNumber);
    state.recent_choices.push(choice.choice_id);
    if (state.recent_choices.length > 8) {
      state.recent_choices = state.recent_choices.slice(-8);
    }
    state.worldline_shift +=
      Math.abs(choice.delta_satisfaction) + Math.abs(choice.delta_clue_points);

    const triggered = [
      ...applyHomeworkProgress(state, choice.choice_id),
      ...applyCrewSync(state, choice.choice_id),
      ...evaluateEvents(state, choice),
      ...evaluateClosedSpaceStage(state, choice.choice_id),
      ...maybeTriggerHomeworkPressure(state),
    ];

    injectResidueBonus(state);
    const eventLabels: string[] = [];
    for (const event of triggered) {
      state.satisfaction = clamp(state.satisfaction + event.delta_satisfaction);
      state.stability = clamp(state.stability + event.delta_stability);
      state.clue_points = clamp(state.clue_points + event.delta_clue_points);
      for (const f of event.add_flags) state.flags.add(f);
      eventLabels.push(formatEventLine(event));
    }

    const ending = evaluateEnding(state);
    if (ending !== null) {
      state.ending_id = ending.ending_id;
      state.ending_title = ending.title;
      state.ending_epilogue = ending.description;
    }

    this._advanceTime(state);

    const record: StepRecord = {
      step_number: stepNumber,
      day: Number(before["day"]),
      timeslot: before["timeslot"] as StepRecord["timeslot"],
      scene_id: scene.scene_id,
      scene_label: scene.label,
      choice_id: choice.choice_id,
      choice_label: choice.label,
      before,
      after: snapshot(state),
      events: eventLabels,
      mutation_profile: { ...profile },
      ending_id: state.ending_id,
      action_flavor: actionFlavor,
    };
    return {
      state,
      record,
      scene,
      choice,
      events: eventLabels,
      ending,
    };
  }

  private _computeProfile(state: GameState): MutationProfile {
    let seed = state.random_seed;
    if (state.mutator_mode === "ai" && seed !== null && seed !== undefined) {
      seed = seed + state.day * 97 + state.loop_count * 193;
    }
    const mutator = buildMutator(state.mutator_mode, seed, state.ai_temperature);
    try {
      return validateProfile(mutator.mutate(state));
    } catch {
      const fallback = buildMutator("deterministic");
      return validateProfile(fallback.mutate(state));
    }
  }

  private _updateStreak(state: GameState, choiceId: string): void {
    if (state.previous_choice === choiceId) {
      state.current_choice_streak += 1;
    } else {
      state.current_choice_streak = 0;
    }
    state.previous_choice = choiceId;
  }

  private _applyRouteUpdates(state: GameState, choice: SceneChoice): void {
    const rs = state.route_state;
    for (const [routeId, delta] of Object.entries(choice.route_progress)) {
      rs.route_progress[routeId] = (rs.route_progress[routeId] ?? 0) + delta;
    }
    for (const [charId, delta] of Object.entries(choice.affinity_delta)) {
      if (charId in rs.character_affinity) {
        rs.character_affinity[charId] = clamp((rs.character_affinity[charId] ?? 0) + delta);
      }
    }
    if (choice.tags.includes("high_risk")) {
      rs.route_tension = Math.max(0, Math.min(20, rs.route_tension + 2));
    } else if (choice.tags.includes("routine") || choice.tags.includes("nagato_relief")) {
      rs.route_tension = Math.max(0, Math.min(20, rs.route_tension - 1));
    }
  }

  private _advanceTime(state: GameState): void {
    if (state.timeslot_index === TIMESLOTS.length - 1) {
      applyMemoryResidue(state);
      state.timeslot_index = 0;
      state.day += 1;
      state.loop_count += 1;
      state.satisfaction = clamp(state.satisfaction - 1);
      state.stability = clamp(state.stability - 1);
      applyResidueDecay(state);
      state.worldline_mutation_profile = this._computeProfile(state);
    } else {
      state.timeslot_index += 1;
    }
    void timeslotOf; // 仅为引用，避免未使用警告
  }
}
